File: Hard/2025-11-05-3321-Find-X-Sum-of-All-K-Long-Subarrays-II/solution.js
var xSum = function(nums, k, x) {
    const n = nums.length;
    const result = new Array(n - k + 1);
    
    const getXSum = (subarray) => {
        const count = {};
        for (const num of subarray) {
            count[num] = (count[num] || 0) + 1;
        }
        
        const freqArray = Object.entries(count).map(([num, cnt]) => ({
            num: parseInt(num),
            cnt
        }));
        
        freqArray.sort((a, b) => b.cnt - a.cnt || b.num - a.num);
        
        const topX = freqArray.slice(0, x);
        const sum = topX.reduce((acc, { num, cnt }) => acc + num * cnt, 0);
        
        return sum;
    };
    
    for (let i = 0; i <= n - k; i++) {
        const subarray = nums.slice(i, i + k);
        result[i] = getXSum(subarray);
    }
    
    return result;
};

File: Hard/2025-11-05-3321-Find-X-Sum-of-All-K-Long-Subarrays-II/solution.py
from collections import Counter
from typing import List

class Solution:
    def xSum(self, subarray: List[int], x: int) -> int:
        count = Counter(subarray)
        most_common = count.most_common(x)
        return sum(val for num, val in most_common)

    def xSumOfAllKLongSubarrays(self, nums: List[int], k: int, x: int) -> List[int]:
        n = len(nums)
        if k == x:
            return [sum(nums[i:i + k]) for i in range(n - k + 1)]
        
        result = []
        current_window = nums[:k]
        result.append(self.xSum(current_window, x))
        
        for i in range(1, n - k + 1):
            current_window = current_window[1:] + [nums[i + k - 1]]
            result.append(self.xSum(current_window, x))
        
        return result

# Example usage:
# sol = Solution()
# print(sol.xSumOfAllKLongSubarrays([1,1,2,2,3,4,2,3], 6, 2))
# print(sol.xSumOfAllKLongSubarrays([3,8,7,8,7,5], 2, 2))